### ISSUE TITLE
Abstract analysis of assume-only contracts

### ISSUE THREAD

**Issue by xaaronc** [kind2-mc/kind2#645]
Suppose I have the following function:
```
function clamp(lo, hi, x: int) returns (y:int)
let
    y = if x < lo then lo
        else if x > hi then hi
        else x;
    --%PROPERTY y >= lo and y <= hi;
tel 
```
The property is reasonable but false because the function is incorrect when `lo > hi`. So I add a precondition with the following contract:
```
(*@contract
    assume lo < hi;
*)
```
Now the property is true, and I can use compositional analysis to verify all callers respect the precondition.

A side-effect is that the function will be abstracted by its contract when analyzing callers of `clamp`. Since there are no `guarantee` clauses in its contract, it receives the default `guarantee true`. For properties dependent on `clamp`, this (likely) leads to failure and a refinement step, impacting verification time.

Is there any reason to attempt an abstract analysis of a contract with only `assume` clauses? If no, is it possible to concretize such nodes before analysis?

**Comment by xaaronc** [comment]
Motivation for the question is I’ve hit some cases where a property is false, but computationally hard to refute, under the default contract, but the property is cheap to prove using the concretized node.

**Comment by daniel-larraz** [comment]
Yes, there is at least one reason. You may want to check whether the satisfaction of a property depends on the definition of `clamp` or not.

If property `y >= lo and y <= hi` is enough for proving properties satisfied by the callers of `clamp`, I suggest to use the following contract:

```
(*@contract
  assume lo < hi;
  guarantee y >= lo and y <= hi;
*)
```

Otherwise, you can achieve your goal by defining `clamp` as an [imported](https://kind.cs.uiowa.edu/kind2_user_doc/2_input/1_lustre.html#the-imported-keyword) function (a function without body) and moving its definition to a contract:

```
function imported clamp(lo, hi, x: int) returns (y:int);
(*@contract
  assume lo < hi;
  guarantee y = (if x < lo then lo else if x > hi then hi else x);
*)
```

Kind 2 will always abstract the imported function by its contract in the analysis.

A final note. The current implementation of IC3 in Kind 2 has some limitations that forces the IC3 engine to stop working when analyzing compositionally Lustre models that includes _functions_ with contracts, or when analyzing models with imported _functions_. If you don’t mind to lose precision in the returned counterexamples by allowing traces where `clamp` does not behave like a function, I recommend defining `clamp` as a _node_ instead of as a function to improve the performance of Kind 2.

**Comment by daniel-larraz** [comment]
Closing this issue. Please feel free to open a new one if you have more questions.

**Comment by daniel-larraz** [comment]
Hi Aaron,

We have reconsidered your suggestion. The expected behavior seems reasonable, and it does not prevent a potential user from forcing an abstract analysis if that is the desired behavior (the user just need to explicitly include `guarantee true` in the contract of the node).

Pull request #730 addresses this change.

Best,
Daniel

**Comment by xaaronc** [comment]
Awesome, thanks a bunch @daniel-larraz. This will be very useful.
